### PR TITLE
Adding QuickSort Algorithm

### LINKED ISSUE #273: Quick Sort
#### Description of the problem
I suggest adding quicksort algorithm in linear data structures
i.e linear_data_structures.algorithms.quick_sort()

There are some sorting algorithms already there, with same nlogn complexity, but since most inbuilt sorting algorithms are implemented using quicksort, we should also add quick sort

### PR DESCRIPTION
**Pull request by anirudh-kac**
Fixes #273 

#### Brief description of what is fixed or changed
I implemented the quick sort algorithm . Added required _partition function and quick_sort function in pydatastructs.linear_data_structures.algorithms
I checked all export statements and init files to make sure it's working

#### Other comments
This is the non parallel version.
I'm also working on the parallel version , but I was facing some issues with that. 
Since some code from this will be reused and this is working fine , I'm making a PR for this version first

### CODE REVIEW

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -360,3 +361,73 @@ def matrix_multiply_parallel(matrix_1, matrix_2, num_threads):
...
+    .. [1] https://en.wikipedia.org/wiki/Quicksort
+    .. [2] https://www.geeksforgeeks.org/quick-sort
+    """
+    start = kwargs.get("start",0)
+    end = kwargs.get("end",len(array)-1)
+    comp = kwargs.get("comp",lambda u,v:u<=v)
+    if start<end:
+        s = _partition(array,start,end,comp)
+        quick_sort(array,start=start,end=s)
+        quick_sort(array,start=s+1,end=end)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:433]: This is recursive. Please convert it into iterative logic.

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:433]: You can use the notion of stacks to do it.

**Comment by czgdp1807** [comment]
Are you participating through GSSoC, 2020?

**Comment by czgdp1807** [comment]
> Since some code from this will be reused

Where can this code be used other than in quick sort algorithm? Serial version isn't needed, only parallel version is to be added. Do not close this PR, just update this with the parallel version of quick sort and remove the sequential version.

**Comment by anirudh-kac** [comment]
@czgdp1807 , I tried using Stack from miscellaneous_data_structures in my implementation of iterative quick sort , however I'm getting import error when importing Stack
` cannot import name 'DynamicOneDimensionalArray' from 'pydatastructs.linear_data_structures'`

**Comment by anirudh-kac** [comment]
Yes I'm participating through GSSOC as of now

**Comment by czgdp1807** [comment]
> I tried using Stack from miscellaneous_data_structures

You can use python list as a stack. No problems.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -360,3 +361,73 @@ def matrix_multiply_parallel(matrix_1, matrix_2, num_threads):
...
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> quick_sort(arr)
+    >>> [arr[0], arr[1], arr[2]]
+    [1, 2, 3]
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Quicksort
+    .. [2] https://www.geeksforgeeks.org/quick-sort
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:425]:
```suggestion
```
Please use, wikipedia/university lectures/research papers as references.

**Comment by anirudh-kac** [comment]
Sorry for the delay, couldn't work for some time.
I removed the recursive logic but how can this be made parallel now? We have a single stack that will be accessed ,won't that be an issue? Can you guide me a little

**Comment by czgdp1807** [comment]
Well, a stack is a list. So you can access any element from inside it. Launch threads for some of the top elements. It's just like a tree, all the leafs are processed parallely in the bottom layer, then the second last, and so on till we reach the top.

**Comment by czgdp1807** [comment]
I think you forgot to add tests for your path. See, tests folder under linear_data_structures.

**Comment by czgdp1807** [comment]
@anirudh-kac Are you working on it? Please do not close the PR.

**Comment by anirudh-kac** [comment]
@czgdp1807  sorry for replying late. I was having some trouble with the parallel implementation and got busy with some college stuff later tbh.
Any resource that might be helpful for implementing the parallel version?

**Comment by czgdp1807** [comment]
See, https://en.wikipedia.org/wiki/Quicksort#Parallelization and a lot PDFs on the web.

**Comment by robotjellyzone** [comment]
Hi @anirudh-kac are you still working on it ! feel free to ask your queries here

**Comment by anirudh-kac** [comment]
@robotjellyzone  I had some trouble implementing the iterative parallel version of quick sort. I went through some resources over the internet but couldn't really figure out how to implement it. I'm still trying but I won't mind if anyone else takes over the issue as I've spent a lot of time

**Comment by robotjellyzone** [comment]
> @robotjellyzone I had some trouble implementing the iterative parallel version of quick sort. I went through some resources over the internet but couldn't really figure out how to implement it. I'm still trying but I won't mind if anyone else takes over the issue as I've taken a lot of time

have you checked this https://en.wikipedia.org/wiki/Quicksort#Parallelization already?

**Comment by Arvind-raj06** [comment]
Is this PR still in progress or should I take it over!

**Comment by sidhu1012** [comment]
> Is this PR still in progress or should I take it over!

You can take it over.

**Comment by czgdp1807** [comment]
Closing in favor of https://github.com/codezonediitj/pydatastructs/pull/314

**Pull request closed without merging**